### PR TITLE
Controller extensibility 

### PR DESCRIPTION
**Pull request by davidkline-ms**
This is to enable easy review of the WIP changes to improve controller extensibility support.

### CODE REVIEW

File: Assets/MRTK/Core/Definitions/Devices/GenericOpenVRControllerDefinition.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    public class GenericOpenVRControllerDefinition : BaseControllerDefinition
+    {
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="handedness"></param>
+        public GenericOpenVRControllerDefinition(Handedness handedness) : base(handedness)
+        { }
+
+        // todo
```
Comment (davidkline-ms) [inline Assets/MRTK/Core/Definitions/Devices/GenericOpenVRControllerDefinition.cs:23]: dont forget to finish this!

File: Assets/MRTK/Core/Definitions/Devices/MixedRealityControllerMapping.cs [hunk 2]
```diff
@@ -221,12 +221,15 @@ private MixedRealityInteractionMapping[] GetDefaultInteractionMappings()
             {
                 switch (handedness)
                 {
+// todo: replace this with the controller definition
```
Comment (davidkline-ms) [inline Assets/MRTK/Core/Definitions/Devices/MixedRealityControllerMapping.cs:224]: dont forget

File: Assets/MRTK/Core/Definitions/Devices/MouseControllerDefinition.cs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    public class MouseControllerDefinition : BaseControllerDefinition
+    {
+        /// <summary>
+        /// 
+        /// </summary>
+        public MouseControllerDefinition() : base(Handedness.None)
+        { }
+
+        /// <inheritdoc />
+        public override MixedRealityInteractionMapping[] DefaultInteractions { get; } =
```
Comment (davidkline-ms) [inline Assets/MRTK/Core/Definitions/Devices/MouseControllerDefinition.cs:23]: => new[] instead

File: Assets/MRTK/Core/Definitions/Devices/GGVHandControllerDefinition.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    /// <summary>
+    /// 
+    /// </summary>
+    public class GGVHandControllerDefinition : BaseControllerDefinition
+    {
+        public GGVHandControllerDefinition() : base(Handedness.None)
+        { }
+
+        /// <inheritdoc />
+        public override MixedRealityInteractionMapping[] DefaultInteractions { get; } = new[]
```
Comment (davidkline-ms) [inline Assets/MRTK/Core/Definitions/Devices/GGVHandControllerDefinition.cs:20]: => new[] instead

File: Assets/MRTK/Core/Definitions/Devices/OculusRemoteControllerDefinition.cs [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+namespace Microsoft.MixedReality.Toolkit.Input
+{
+    /// <summary>
+    /// 
+    /// </summary>
+    public class OculusRemoteControllerDefinition : BaseControllerDefinition
+    {
+        /// <summary>
+        ///
```
Comment (davidkline-ms) [inline Assets/MRTK/Core/Definitions/Devices/OculusRemoteControllerDefinition.cs:17]: todo: missing doc

File: Assets/MRTK/Core/Definitions/Devices/MouseControllerDefinition.cs [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+namespace Microsoft.MixedReality.Toolkit.Input
+{
+    /// <summary>
+    /// Defines the base interactions and data that an controller can provide.
+    /// </summary>
+    public class MouseControllerDefinition : BaseControllerDefinition
+    {
+        /// <summary>
+        ///
```
Comment (davidkline-ms) [inline Assets/MRTK/Core/Definitions/Devices/MouseControllerDefinition.cs:17]: todo: missing comment

File: Assets/MRTK/Core/Definitions/Devices/TouchScreenControllerDefinition.cs [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    public class TouchScreenControllerDefinition : BaseControllerDefinition
+    {
+        /// <summary>
+        /// 
+        /// </summary>
+        public TouchScreenControllerDefinition() : base(Handedness.None)
+        { }
+
+        /// <inheritdoc />
+        public override MixedRealityInteractionMapping[] DefaultInteractions { get; } =
```
Comment (davidkline-ms) [inline Assets/MRTK/Core/Definitions/Devices/TouchScreenControllerDefinition.cs:23]: => new[]

File: Assets/MRTK/Core/Definitions/Devices/WindowsMixedRealityControllerDefinition.cs [hunk 8]
```diff
@@ -36,7 +35,7 @@ public WindowsMixedRealityControllerDefinition(IMixedRealityInputSource source,
             new MixedRealityInteractionMapping(8, "Touchpad Press", AxisType.Digital, DeviceInputType.TouchpadPress),
             new MixedRealityInteractionMapping(9, "Menu Press", AxisType.Digital, DeviceInputType.Menu),
             new MixedRealityInteractionMapping(10, "Thumbstick Position", AxisType.DualAxis, DeviceInputType.ThumbStick),
-            new MixedRealityInteractionMapping(11, "Thumbstick Press", AxisType.Digital, DeviceInputType.ThumbStickPress)
```
Comment (davidkline-ms) [inline Assets/MRTK/Core/Definitions/Devices/WindowsMixedRealityControllerDefinition.cs:39]: why did this change?

File: Assets/MRTK/Core/Definitions/Devices/WindowsMixedRealityControllerDefinition.cs [hunk 9]
```diff
@@ -2,32 +2,31 @@
...
+        /// <param name="handedness"></param>
+        public WindowsMixedRealityControllerDefinition(Handedness handedness) : base(handedness)
+        { }
+
+        /// <inheritdoc />
+        public override MixedRealityInteractionMapping[] DefaultInteractions => new[]
         {
             new MixedRealityInteractionMapping(0, "Spatial Pointer", AxisType.SixDof, DeviceInputType.SpatialPointer),
             new MixedRealityInteractionMapping(1, "Spatial Grip", AxisType.SixDof, DeviceInputType.SpatialGrip),
-            new MixedRealityInteractionMapping(2, "Grip Press", AxisType.SingleAxis, DeviceInputType.GripPress),
```
Comment (davidkline-ms) [inline Assets/MRTK/Core/Definitions/Devices/WindowsMixedRealityControllerDefinition.cs:30]: why did this change?

Comment (davidkline-ms) [reply Assets/MRTK/Core/Definitions/Devices/WindowsMixedRealityControllerDefinition.cs:30]: should be grippress

File: Assets/MRTK/Core/Definitions/Devices/WindowsMixedRealityControllerDefinition.cs (hunk 8 above)
Comment (davidkline-ms) [reply Assets/MRTK/Core/Definitions/Devices/WindowsMixedRealityControllerDefinition.cs:39]: remove the comma

File: Assets/MRTK/Providers/LeapMotion/LeapMotionArticulatedHand.cs [hunk 10]
```diff
@@ -31,15 +31,18 @@ public class LeapMotionArticulatedHand : BaseHand
...
-            : base(trackingState, controllerHandedness, inputSource, interactions)
-        {
-            handDefinition = new ArticulatedHandDefinition(inputSource, controllerHandedness);
-        }
+        public LeapMotionArticulatedHand(
+            TrackingState trackingState,
+            Handedness controllerHandedness,
+            IMixedRealityInputSource inputSource = null,
+            MixedRealityInteractionMapping[] interactions = null)
+                : base(trackingState, controllerHandedness, inputSource, interactions) // todo: controller definition here
```
Comment (davidkline-ms) [inline Assets/MRTK/Providers/LeapMotion/LeapMotionArticulatedHand.cs:39]: dont forget

File: Assets/MRTK/Providers/OpenVR/WindowsMixedRealityOpenVRMotionController.cs [hunk 11]
```diff
@@ -19,10 +20,14 @@ public class WindowsMixedRealityOpenVRMotionController : GenericOpenVRController
...
-            : base(trackingState, controllerHandedness, inputSource, interactions)
+        public WindowsMixedRealityOpenVRMotionController(
+            TrackingState trackingState,
+            Handedness controllerHandedness,
+            IMixedRealityInputSource inputSource = null,
+            MixedRealityInteractionMapping[] interactions = null)
+                : base(trackingState, controllerHandedness, inputSource, interactions, new WindowsMixedRealityControllerDefinition(controllerHandedness))
         {
-            controllerDefinition = new WindowsMixedRealityControllerDefinition(inputSource, controllerHandedness);
+            controllerDefinition = new WindowsMixedRealityControllerDefinition(controllerHandedness);
```
Comment (davidkline-ms) [inline Assets/MRTK/Providers/OpenVR/WindowsMixedRealityOpenVRMotionController.cs:30]: remove me

File: Assets/MRTK/Providers/OpenVR/WindowsMixedRealityOpenVRMotionController.cs [hunk 12]
```diff
@@ -19,10 +20,14 @@ public class WindowsMixedRealityOpenVRMotionController : GenericOpenVRController
...
         }
 
         private readonly WindowsMixedRealityControllerDefinition controllerDefinition;
```
Comment (keveleigh) [inline Assets/MRTK/Providers/OpenVR/WindowsMixedRealityOpenVRMotionController.cs:33]: Update this to point to base definition

File: Assets/MRTK/Providers/WindowsMixedReality/XR2018/Controllers/WindowsMixedRealityArticulatedHand.cs [hunk 13]
```diff
@@ -35,8 +35,16 @@ public class WindowsMixedRealityArticulatedHand : BaseWindowsMixedRealitySource,
...
+            Handedness controllerHandedness,
+            IMixedRealityInputSource inputSource = null,
+            MixedRealityInteractionMapping[] interactions = null)
+                : base(trackingState,
+                      controllerHandedness,
+                      new ArticulatedHandDefinition(inputSource, controllerHandedness),
+                      inputSource,
+                      interactions)
         {
             handDefinition = new ArticulatedHandDefinition(inputSource, controllerHandedness);
```
Comment (davidkline-ms) [inline Assets/MRTK/Providers/WindowsMixedReality/XR2018/Controllers/WindowsMixedRealityArticulatedHand.cs:49]: remove

File: Assets/MRTK/Providers/WindowsMixedReality/XR2018/Controllers/WindowsMixedRealityArticulatedHand.cs [hunk 14]
```diff
@@ -46,12 +54,6 @@ public WindowsMixedRealityArticulatedHand(TrackingState trackingState, Handednes
                 "TryGetHandPose");
         }
 
-        /// <summary>
-        /// The Windows Mixed Reality articulated hands default interactions.
-        /// </summary>
-        /// <remarks>A single interaction mapping works for both left and right articulated hands.</remarks>
-        public override MixedRealityInteractionMapping[] DefaultInteractions => handDefinition?.DefaultInteractions;
```
Comment (keveleigh) [inline Assets/MRTK/Providers/WindowsMixedReality/XR2018/Controllers/WindowsMixedRealityArticulatedHand.cs:53]: Should this be left in but obsolete?

**Review by davidkline-ms — COMMENTED** [review]
a few comments

**Pull request closed without merging**